---
title: "PopVar Post-Hoc Predictions"
output: html_notebook
---

## Introduction

Predictions were made via PopVar using genotypic information from the previous
reference, and potentially uncorrected phenotypic data. Here I use new genotypic
data and correct phenotypic data to make predictions of the crosses that 
were already made.

```{r}

library(tidyverse)
library(PopVar)
library(readxl)
library(stringr)


geno.file <- "C:/Users/Jeff/Google Drive/Barley Lab/Projects/Genomic Selection/Genotypic Data/GBS Genotype Data/S2_genos_hmp.RData"
pheno.file <- "C:/Users/Jeff/Google Drive/Barley Lab/Projects/Genomic Selection/Phenotypic Data/Final/Master Phenotypes/PopVarVal_BLUE.RData"

# Cross file
cross.file <- "C:/Users/Jeff/Google Drive/Barley Lab/Projects/PopVarVal/Crosses/S2_F15_crosses_Jeff.xlsx"

# Read in data
load(geno.file)
load(pheno.file)

crosses <- read_excel(cross.file) %>%
  # Remove some columns 
  subset(select = -7:-9) %>%
  rename(Par1 = `Parent 1`, Par2 = `Parent 2`, Experiment = `What experiment?`, Success = `Was the cross successful?`)

```


## Preparation

Need to manipulate phenotypic data, map information, and the genotype matrix
for use in PopVar

```{r}

# Training and prediction line names
tp <- names(s2_imputed_genos) %>%
  str_subset("^[0-9]{2}")

vp <- names(s2_imputed_genos) %>%
  str_subset("^2MS")

# Trim the phenotype data.frame
y.in <- PopVarVal.tidy.BLUE %>%
  filter(line_name %in% tp) %>%
  spread(trait, value) %>%
  as.data.frame()

# Edit the snp info for the map
map.in <- s2_imputed_genos %>%
  select(`rs#`, chrom, cM_pos) %>%
  rename(mkr = `rs#`) %>%
  as.data.frame()

# Pre-allocate the G.in data.frame
G.in <- matrix(NA, nrow = length(tp) + length(vp) + 1, ncol = nrow(map.in) + 1) %>%
  as.data.frame()

# Add entry names
G.in[,1] <- c("mkr", tp, vp)
# Add marker names
G.in[1,-1] <- map.in$mkr
# Add genotypic data
G.in[-1, -1] <- t(s2_imputed_genos[,-1:-5]) %>% as.data.frame()


```



## Re-Run Predictions for all traits

```{r, cache=TRUE}

# Generate the crossing table from the cross file
crossing.table <- crosses %>% select(Par1, Par2)

# Run simulations
popvar.out <- pop.predict(G.in = G.in, y.in = y.in, map.in = map.in, 
                          crossing.table = crossing.table, nInd = 150,
                          nSim = 10, min.maf = 0, mkr.cutoff = 1, entry.cutoff = 1,
                          nCV.iter = 0, impute = "pass", models = "rrBLUP")

# Save
save.file <- "Predictions/post_hoc_predictions_out.RData"
save("popvar.out", file = save.file)

```


## Visualize

```{r}

load(save.file)

# For each trait extract the mean, V_G and mu_sp
pred.details <- lapply(X = names(popvar.out$predictions), FUN = function(trait) {
  trait.name <- trait %>% str_replace("_param.df", "")
  popvar.out$predictions[[trait]] %>% 
    as.data.frame() %>% 
    # Need to unlist everything
    mutate_each(funs = funs(unlist)) %>%
    mutate(trait = trait.name) %>%
    select(Par1, Par2, trait, pred.mu, pred.varG, mu.sp_low, mu.sp_high) }) %>%
  bind_rows()

# Combine with crossing table
crosses1 <- full_join(crosses, pred.details) %>%
  # Sort by Experiment
  arrange(trait, Experiment)

# Now lets plot mean verusus predicted variance
crosses1 %>%
  filter(!is.na(trait)) %>%
  ggplot(aes(x = pred.mu, y = pred.varG)) +
  geom_point() +
  xlab("Predicted Mean") +
  ylab(expression(Predicted~V[G])) +
  facet_wrap("trait", scales = "free")

# Save
save.file <- "Figures/post_hoc_mean_versus_var.jpg"
ggsave(save.file, height = 6, width = 8)


# Find the fold difference between the highest predicted V_G and 
# the lowest predicted V_G for each trait
crosses1 %>%
  filter(!is.na(trait)) %>%
  # Create a cross column
  mutate(cross = str_c(Par1, Par2, sep = "/")) %>%
  group_by(trait) %>% 
  summarize(min.V_G = min(pred.varG),
            cross.min.V_G = cross[which.min(pred.varG)],
            max.V_G = max(pred.varG),
            cross.max.V_G = cross[which.max(pred.varG)]) %>% 
  mutate(fold.diff = max.V_G / min.V_G)


```



